drivers: display: hub12: add horizontal chaining support

heronet opened the pull request:
- Add support for chaining multiple 32x16 panels horizontally by setting width to multiples of 32 (64, 96, 128, etc).
- Fix a minor race condition issue.

heronet commented on drivers/display/display_hub12.c:
```diff
@@ -142,7 +160,7 @@ static int hub12_write(const struct device *dev, const uint16_t x, const uint16_
 	}
 
-	k_sem_take(&data->lock, K_FOREVER);
+	k_mutex_lock(&data->lock, K_FOREVER);
```
Mutex seemed like a better fit since we're protecting shared data access. Also helps if multiple threads end up calling display_write at different priorities

JarmouniA replied:
Expect for priority inversion mitigation, a semaphore is lighter than & recommended over a mutex in Zephyr. A mutex makes the function unusable from an ISR.
Also calling display_write from multiple threads would be a bad practice in my opinion.

https://academy.nordicsemi.com/courses/nrf-connect-sdk-fundamentals/lessons/lesson-8-thread-synchronization/topic/mutexes/
https://academy.nordicsemi.com/courses/nrf-connect-sdk-fundamentals/lessons/lesson-8-thread-synchronization/topic/semaphores/

heronet committed:
dts: bindings: display: hub12: add horizontal chaining support

- Add support for chaining multiple 32x16 panels horizontally by
  setting width to multiples of 32 (64, 96, 128, etc).

Signed-off-by: Siratul Islam <email@sirat.me>

rruuaanng reviewed (commented):
Looks good

JarmouniA commented on drivers/display/display_hub12.c:
```diff
@@ -39,11 +40,13 @@ struct hub12_config {
 	struct spi_dt_spec spi;
 	uint16_t width;
 	uint16_t height;
+	uint8_t num_panels;
```
Should be last in the struct.

JarmouniA replied:
> I didn't realize this was hanging for 2 weeks. Done

no it wasn't, I forgot to submit it.

heronet commented on drivers/display/display_hub12.c:
```diff
@@ -97,16 +109,23 @@ static void hub12_scan_row(struct hub12_data *data, const struct hub12_config *c
+
+		hub12_update_cache(data, config, next_row);
+
+		k_sem_give(&data->lock);
```
This is intentional

heronet commented on dts/bindings/display/zephyr,hub12.yaml:
```diff
@@ -45,11 +49,8 @@ properties:
     description: |
-      Display width in pixels. Must match the physical hardware panel.
-      Currently only single 32-pixel wide panels are supported.
+      Display width in pixels. Must be a multiple of 32 for horizontal chaining.
```
Already specified in height property

heronet commented on drivers/display/display_hub12.c:
```diff
@@ -97,16 +109,23 @@ static void hub12_scan_row(struct hub12_data *data, const struct hub12_config *c
+
+		hub12_update_cache(data, config, next_row);
+
+		k_sem_give(&data->lock);
```
This is intentional

heronet committed:
drivers: display: hub12: add horizontal chaining support

- Add support for chaining multiple 32x16 panels horizontally
- Also fixed a race condition issue.

Signed-off-by: Siratul Islam <email@sirat.me>
